src/cryptsetup/cryptsetup-tokens/cryptsetup-token-systemd-tpm2.c

```diff
@@ -77,6 +68,17 @@ _public_ int cryptsetup_token_open(
         assert(token == r);
         assert(json);
 
+        if (pin && memchr(pin, 0, pin_size - 1))
+                return crypt_log_error_errno(cd, ENOANO, "PIN must be characters string.");
+
+        /* pin was passed as pin = pin, pin_size = strlen(pin). We need to add terminating
+         * NULL byte to addressable memory */
+        if (pin && pin[pin_size-1] != '\0') {
+                pin_string = memdup_suffix0(pin, pin_size);
+                if (!pin_string)
+                        return crypt_log_oom(cd);
+        }
```

Comment:
I'd simplify this in two ways: merge the two if blocks, and simply always do the memdup_suffix0(). it's not worth optimizing here. simple codeflow we always value the most. hence maybe:

```c
if (pin) {
        if (memchr(…)) /* Refuse embedded NUL, allow trailing NUL */
                return …

        pin_string = memdup_suffix0(pin, pin_size); /* Enforce trailing NUL if missing */
        if (!pin_string)
                return …
}
```

and then get rid of the ternary op when using pin_string.

Reply:
Implemented by introducing a new `crypt_normalize_pin()` function as suggested below and dropping the ternary operator as part of the refactoring.